src/twisted/enterprise/adbapi.py

```diff
@@ -463,18 +463,44 @@ def _close(self, conn):
     def _runInteraction(self, interaction, *args, **kw):
         conn = self.connectionFactory(self)
         trans = self.transactionFactory(self, conn)
-        try:
-            result = interaction(trans, *args, **kw)
-            trans.close()
-            conn.commit()
-            return result
-        except:
-            excType, excValue, excTraceback = sys.exc_info()
+        if self.reconnect and self.dbapiName == 'MySQLdb':
+            from MySQLdb import OperationalError
             try:
-                conn.rollback()
+                result = interaction(trans, *args, **kw)
+                trans.close()
+                conn.commit()
+                return result
+            except OperationalError as e:
+                # 2006: MySQL server has gone away
+                # 2013: Lost connection to MySQL server
+                # 1213: Deadlock found when trying to get lock
+                if e[0] not in (2006, 2013, 1213):
```

Comment:
This is another problem: different MySQL versions might have different error codes that indicate an action could be retried.  It'd be hard to keep this list up to date as well.